src/types.py

```diff
@@ -117,64 +132,141 @@ def is_int(self) -> bool:
+        return self.get_representative().kind == Type.K_FN
+
+    def is_void(self) -> bool:
+        return self.get_representative().kind == Type.K_VOID
+
     def is_unsigned(self) -> bool:
         return self.get_representative().sign == Type.UNSIGNED
 
     def get_size_bits(self) -> int:
         return self.get_representative().size or 32
 
+    def get_size_align_bytes(self) -> Tuple[int, int]:
+        type = self.get_representative()
+        if type.is_ctype() and type.ctype_ref is not None:
+            assert type.typemap is not None
+            return var_size_align(type.ctype_ref, type.typemap)
+        size = type.get_size_bits() // 8
+        return size, size
+
+    def is_struct_type(self) -> bool:
+        type = self.get_representative()
+        if type.kind != Type.K_CTYPE or type.ctype_ref is None:
+            return False
+        assert type.typemap is not None
+        ctype = resolve_typedefs(type.ctype_ref, type.typemap)
+        if not isinstance(ctype, ca.TypeDecl):
+            return False
+        return isinstance(ctype.type, (ca.Struct, ca.Union))
+
+    def get_pointer_target(self) -> Optional["Type"]:
+        """If self is a pointer-to-a-Type, return the Type"""
+        type = self.get_representative()
+        if type.is_pointer() and type.ptr_to is not None:
+            return type.ptr_to
+        return None
+
     def get_pointer_to_ctype(self) -> Optional[CType]:
         """If self is a pointer-to-a-CType, return the CType"""
+        ptr_to = self.get_pointer_target()
+        if ptr_to is not None and ptr_to.is_ctype() and ptr_to.ctype_ref:
+            return ptr_to.ctype_ref
+        return None
+
+    def get_function_pointer_signature(self) -> Optional["FunctionSignature"]:
+        """If self is a function pointer, return the FunctionSignature"""
         type = self.get_representative()
         if type.is_pointer() and type.ptr_to is not None:
             ptr_to = type.ptr_to.get_representative()
-            if ptr_to.is_ctype() and ptr_to.ctype_ref:
-                return ptr_to.ctype_ref
+            if ptr_to.kind == Type.K_FN:
+                return ptr_to.fn_sig
         return None
 
-    def parse_function(self) -> Optional[CFunction]:
-        ctype = self.get_pointer_to_ctype()
-        if ctype is not None:
-            return parse_function(ctype)
-        return None
+    def to_decl(self, name: str, fmt: Formatter) -> str:
+        decl = ca.Decl(
+            name=name,
+            type=self._to_ctype(set(), fmt),
+            quals=[],
+            storage=[],
+            funcspec=[],
+            init=None,
+            bitsize=None,
+        )
+        set_decl_name(decl)
+        return to_c(decl)
 
-    def to_decl(self, var: str, fmt: Formatter) -> str:
-        ret = self.format(fmt)
-        prefix = ret if ret.endswith("*") else ret + " "
-        return prefix + var
+    def _to_ctype(self, seen: Set["Type"], fmt: Formatter) -> CType:
```

Comment:
btw, I wonder if we should use this pattern of creating a pycparser type for printing in more places. Could help with precedence and avoiding excessive parentheses.

Reply:
Yeah it's a good point. I feel like it might make the code look a bit more verbose, but it'd have much fewer branches & be more robust.
I think types are particularly hard to print right, so this seemed like a good place to start.